src/main/java/com/amazon/ion/impl/bin/PresenceBitmap.kt

```diff
@@ -88,22 +117,30 @@ internal class PresenceBitmap(
 
         /** Creates or reuses a [PresenceBitmap] for the given signature. */
         @JvmStatic
-        fun create(signature: List<Parameter>): PresenceBitmap {
+        fun create(signature: List<Parameter>, pool: PooledFactory): PresenceBitmap {
             if (signature.size > MAX_SUPPORTED_PARAMETERS) throw IonException("Macros with more than 128 parameters are not supported by this implementation.")
             // Calculate the actual number of presence bits that will be encoded for the given signature.
-            val nonRequiredParametersCount = signature.count { it.cardinality != ParameterCardinality.ExactlyOne }
-            val usePresenceBits = nonRequiredParametersCount > PRESENCE_BITS_SIZE_THRESHOLD || signature.any { it.type.taglessEncodingKind != null }
+            var nonRequiredParametersCount = 0
+            var usePresenceBits = false
+            for (i in signature.indices) {
+                val parameter = signature[i]
+                if (parameter.cardinality != ParameterCardinality.ExactlyOne) {
+                    nonRequiredParametersCount++
+                }
+                usePresenceBits = usePresenceBits or (parameter.type.taglessEncodingKind != null)
+            }
+            usePresenceBits = usePresenceBits or (nonRequiredParametersCount > PRESENCE_BITS_SIZE_THRESHOLD)
```

Comment:
The result is the same, but the new (uglier) version achieves two things:
1. It eliminates `ArrayList$Itr` allocations that are implicit to Kotlin's `List.count` and `List.any` methods. These were showing up prominently in memory profiles. This change alone lowered allocation rate from 147 KB/op to 139 KB/op.
2. It only loops through the signature once, instead of (worst case) two full times.